processing/src/main/java/org/apache/druid/segment/filter/Filters.java

```diff
@@ -426,175 +424,21 @@ public static Filter convertToCNFFromQueryContext(Query query, @Nullable Filter
       return null;
     }
     boolean useCNF = query.getContextBoolean(CTX_KEY_USE_FILTER_CNF, false);
-    return useCNF ? toCNF(filter) : filter;
+    return useCNF ? Filters.toCnf(filter) : filter;
   }
 
-  public static Filter toCNF(Filter current)
+  public static Filter toCnf(Filter current)
   {
-    current = pushDownNot(current);
-    current = flatten(current);
-    current = convertToCNFInternal(current);
-    current = flatten(current);
+    current = HiveCnfHelper.pushDownNot(current);
+    current = HiveCnfHelper.flatten(current);
+    // Pull out AND filters first to convert the filter into a conjunctive form.
+    // This is important to not create a huge CNF.
+    current = CalciteCnfHelper.pull(current);
+    current = HiveCnfHelper.convertToCNFInternal(current);
+    current = HiveCnfHelper.flatten(current);
```

Comment:
It would be nice to add some commentary on each line describing (possibly with example) why it exists. I know there is something in PR description but this is non-trivial.

Reply:
This code was adopted from Apache Hive as well and I don't know the exact reason for why we are doing in this order.. I tried to add some. Hopefully it helps.

Reply:
thanks, those comments do help.

I think I got/remember it now.

"pushDownNot, flatten, convertToCnf, flatten" sequence is to convert original boolean expression into the CNF form, which doesn't necessarily "optimize"(from evaluation perspective) but CNF form is useful for analysis on boolean expressions(often  use to  prove equivalence of two different boolean expressions), in  our case, separating filter evaluation into pre/post cursor use for applying filters.

your PRs (this and #9608 ) are towards optimizing the boolean expression. Are your PRs driven by similar work done in Calcite or some know work or based  on your own heuristics ? For example there are optimizations like (not sure if they are taken care of or not)
`a OR (NOT  a) = TRUE`
`a AND (NOT A) = FALSE`
... and more listed in section "Ten Basic Rules of Boolean Algebra, in https://grace.bluegrass.kctcs.edu/~kdunn0001/files/Simplification/4_Simplification_print.html

unfortunate thing is that this is a major research topic with no proven algorithm that works in all cases.

BTW: this comment is not a PR code review but just general things I noted while thinking more about this and wanted to document here.

Reply:
Yeah, my two PRs are for creating more optimized CNFs in terms of size since we have seen that a filter of the same pattern with that at https://github.com/apache/druid/pull/9634#issuecomment-611156592 blows up memory after CNF conversion. Those PRs are mostly based on my heuristics. I happened to find `RexUtil.pullFactors()` of Calcite which seems useful to address the size issue. 

> For example there are optimizations like (not sure if they are taken care of or not)
> a OR (NOT a) = TRUE
> a AND (NOT A) = FALSE

Good point. I'm aware of that our CNF conversion doesn't reduce such filters by evaluating them, but I hope Calcite could reduce most of them in the SQL layer. After the join support was added, now there are two main use cases for the filter in CNF. One is cursor filter push down to use bitmaps if possible (the one you mentioned) and another is join filter push down to place filters at proper joins (both happen in historicals). I think it would be safe to assume that the filter is already optimized by Calcite for these use cases since SQL is now the primary way for querying.

> unfortunate thing is that this is a major research topic with no proven algorithm that works in all cases.

I totally agree. @jon-wei and I have been talking about it and perhaps the best would be using a well-tested third-party library. But I haven't seen such a library yet.